# 01. Importing Libraries

In [3]:
# Importing libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [5]:
# Creating a path to main project folder
path = r'C:\Users\jboer\OneDrive\Documents\Career Foundry\Instacart Basket Analysis'

In [6]:
# Importing products.csv file
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'), index_col = False)

In [7]:
# Importing orders_wrangled.csv file
df_ords = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'), index_col = False)

# 03. Creating a test Dataframe

In [8]:
# Create a dataframe
df_test = pd.DataFrame()

In [74]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [75]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [76]:
# check for mixed types in dataframe
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) !=df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


# 04. Missing Values

In [9]:
# Finding missing values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
# Viewing missing values by creating a dataframe for them
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
# Checking number of rows in current dataframe
df_prods.shape

(49693, 5)

In [13]:
# Creating a new dataframe with only "non-missing" values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [14]:
df_prods_clean.shape

(49677, 5)

# 05. Duplicates

In [15]:
# Finding full duplicates in a dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
# Creating new dataframe that doesn't include duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

# 06. Exporting New Dataframe

In [20]:
# Exporting new dataframe as to not lose changes
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# **Beginning of Task**

### 1. Importing New Data

In [2]:
# Importing products_checked.csv file
df_prods = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','products_checked.csv'), index_col = False)

NameError: name 'pd' is not defined

### 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!


In [22]:
# Performing inital check of df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

##### 1. order_number: has min and max values of 1 and 100, respectively, meaning the most orders from a customer is 100.
##### 2. orders_day_of_week: min value is "0" which represents Saturday and max value is "6" which represents Friday. Also, shows there are 7 values in this column, one for every day of the week.
##### 3. order_hour_of_day: min value of "0" represents 12AM and max value of "23" represents 11PM. 24 values in the column = 24 hours in a day.
##### 4. days_since_prior_order: has a count less than all other columns which could mean there are missing values. A min of "0" indicates that someone placed at least 2 orders in the same day and a max of "30" means no customer went longer than 31 days before placing another order (with "0" included there are 31 values). It may also mean that this column's value cap is 30 (1 month) and anything over that won't be recorded. 

### 3. Check for mixed-type data in your df_ords dataframe.

In [23]:
# Checking for mix-type data in df_ords
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) !=df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

In [24]:
# Reviewing data types in df_ords
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

##### No mix-type data was found in df_ords dataframe.

### 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

In [25]:
# Changing data type of "days_since_prior_order" column to "float64"
df_ords['days_since_prior_order'] = df_ords['days_since_prior_order'].astype('float64')

##### After attempting to replace null values in a later step in this exercise, I tried your suggestion of changing the data type to float64 exclusively.

### 5. Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.


In [26]:
# Finding missing values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

##### There were 206,209 missing values found in the "days_since_prior_order" column. An explanation for this could be that those customers never placed another order, so there would be no value to enter in that column.

### 6. Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice

In [27]:
# Creating a dataframe for missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [28]:
# Reviewing missing values dataframe
df_ords_nan.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
11          11   2168274        2             1                   2   
26          26   1374495        3             1                   1   
39          39   3343014        4             1                   6   
45          45   2717275        5             1                   3   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
11                 11                     NaN  
26                 14                     NaN  
39                 11                     NaN  
45                 12                     NaN

In [29]:
# Reviewing missing values dataframe
df_ords_nan.tail()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN

In [30]:
# Creating new dataframe and replacing missing values with "0"
df_ords_nonulls = df_ords.copy()
df_ords_nonulls['days_since_prior_order'] = df_ords_nonulls['days_since_prior_order'].fillna(0)

In [33]:
# Checking new dataframe
df_ords_nonulls.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     0.0  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

##### After getting an error when trying to replace null values with "0", I used your suggestion to change the data type of the "days_since_prior_order" column to float64 exclusivley which worked. The reason I chose to replace the null values with "0" is because I noticed that all the null values in this column were on the first order for every customer. This makes sense why it would be null because it's their first order so there are no number of days since their last order. That's why I chose to replace the null with "0". It is a value that makes sense, there are 0 days since their last order because it is their first order, and imputing "0" won't affect my analysis.

### 7. Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.


In [34]:
# Finding full duplicates in a df_ords dataframe
df_ords_dups = df_ords_nonulls[df_ords_nonulls.duplicated()]

In [38]:
# Checking new dataframe for duplicates
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [39]:
# Finding duplicates in the "order_id" column
df_ords_dups_1 = df_ords_nonulls[df_ords_nonulls['order_id'].duplicated()]

In [37]:
df_ords_dups_1

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

##### No full duplicates were found (the df_ords_dups dataframe). I thought I'd also check for duplicates in the order_id column (the df_ords_dups_1 dataframe) because that is the only column in the dataframe that should have unique values, but there were no duplicates found.

### 8. Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.


##### Since there were no duplicates found, there is no need to address any. But if there were I would use the following sequence of codes:
#####  **Checking row count of df_ords_nonulls**
df_ords_nonulls.shape
##### **Creating new dataframe that doesn't include duplicates**
df_ords_clean = df_ords_nonulls.drop_duplicates()
##### **Checking row count of new dataframe to see if duplicates were removed**
df_ords_clean.shape

### 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [40]:
# Exporting df_ords_nonulls dataframe as to not lose changes
df_ords_nonulls.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

##### The final, cleaned df_prods was exported in the last step of the exercise before the beginning of my task.